In [2]:
#loading clean dataset for exploratory data analysis
import pandas as pd
import numpy as np

df = pd.read_csv("../Data/Processed/loan_clean_dataset.csv")

In [3]:
#dataset anlysis
df.head()

LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0  DSL4O0KAWD   64   73743      140354          300               0   
1  S5IDRA0LCA   22   46351      151652          300               0   
2  4X63ZOH9Y4   56   42359      163692          300               0   
3  7QM3SGA1AU   23   52705      218580          300               0   
4  CDH9OTQ8H6   58  131487       15557          301               0   

   NumCreditLines  InterestRate  LoanTerm  DTIRatio  ... EmploymentType  \
0               2          4.12        12      0.24  ...  Self-employed   
1               4         10.04        36      0.51  ...      Part-time   
2               3          4.98        60      0.71  ...      Part-time   
3               3         22.31        60      0.54  ...     Unemployed   
4               4         21.35        36      0.38  ...      Part-time   

  MaritalStatus HasMortgage HasDependents LoanPurpose HasCoSigner Default  \
0        Single         Yes            No   Education         Yes       0   
1        Single          No           Yes       Other         Yes       0   
2      Divorced          No           Yes   Education          No       0   
3      Divorced          No           Yes    Business         Yes       0   
4      Divorced          No            No       Other         Yes       0   

   hard_invalid_flag  soft_anomaly_flag                     validation_reason  
0                  0                  1           Loan Amount Exceeds Income;  
1                  0                  1           Loan Amount Exceeds Income;  
2                  0                  1  High DTI;Loan Amount Exceeds Income;  
3                  0                  1           Loan Amount Exceeds Income;  
4                  0                  0                                   NaN  

[5 rows x 21 columns]

In [4]:
df.shape

(143440, 21)

In [5]:
df.columns

Index(['LoanID', 'Age', 'Income', 'LoanAmount', 'CreditScore',
       'MonthsEmployed', 'NumCreditLines', 'InterestRate', 'LoanTerm',
       'DTIRatio', 'Education', 'EmploymentType', 'MaritalStatus',
       'HasMortgage', 'HasDependents', 'LoanPurpose', 'HasCoSigner', 'Default',
       'hard_invalid_flag', 'soft_anomaly_flag', 'validation_reason'],
      dtype='object')

In [6]:
#checking null value
df.isnull().count()

LoanID               143440
Age                  143440
Income               143440
LoanAmount           143440
CreditScore          143440
MonthsEmployed       143440
NumCreditLines       143440
InterestRate         143440
LoanTerm             143440
DTIRatio             143440
Education            143440
EmploymentType       143440
MaritalStatus        143440
HasMortgage          143440
HasDependents        143440
LoanPurpose          143440
HasCoSigner          143440
Default              143440
hard_invalid_flag    143440
soft_anomaly_flag    143440
validation_reason    143440
dtype: int64

In [7]:
#Task1
#total number of loans

total_no_of_loans= df["LoanID"].count()
total_no_of_loans

np.int64(143440)

In [8]:
#total number of defaults

total_no_of_defaults =(df["Default"]==1).sum()
total_no_of_defaults

np.int64(15355)

In [9]:
#default rate
default_rate= total_no_of_defaults/total_no_of_loans*100
default_rate = round(default_rate, 2)
default_rate

np.float64(10.7)

In [10]:
#Task 2 
#default comparison for each feature

features = ["CreditScore","Income","LoanAmount","InterestRate","DTIRatio","MonthsEmployed"]

mean = df.groupby("Default")[features].mean().round(2)
median = df.groupby("Default")[features].median().round(2)
print("--------mean value for each feature-------")
print(mean)
print("--------median value for each feature-------")
print(median)


--------mean value for each feature-------
         CreditScore    Income  LoanAmount  InterestRate  DTIRatio  \
Default                                                              
0             575.65  83901.07   125317.28         13.18      0.50   
1             558.95  71847.64   144858.32         15.85      0.51   

         MonthsEmployed  
Default                  
0                 60.00  
1                 48.78  
--------median value for each feature-------
         CreditScore   Income  LoanAmount  InterestRate  DTIRatio  \
Default                                                             
0              576.0  84279.0    124162.0         13.00      0.50   
1              552.0  66560.0    153486.0         16.86      0.52   

         MonthsEmployed  
Default                  
0                  60.0  
1                  44.0  


In [11]:
#Task 3
#Default rate analysis for Employment type

summary = df.groupby("EmploymentType")["Default"].agg(
    total_loans="count",
    defaults="sum",
    default_rate="mean"
).round(3)

print(summary)

                total_loans  defaults  default_rate
EmploymentType                                     
Full-time             47644      4450         0.093
Part-time             47891      5557         0.116
Self-employed         47501      5261         0.111
Unemployed              404        87         0.215


In [12]:
#Default rate analysis for Education
summary = df.groupby("Education")["Default"].agg(
    total_loans="count",
    defaults="sum",
    default_rate="mean"
).round(3)

print(summary)

             total_loans  defaults  default_rate
Education                                       
Bachelor's         36240      4145         0.114
High School        35855      4207         0.117
Master's           35753      3553         0.099
PhD                35592      3450         0.097


In [13]:
#Default rate analysis for Marital Status
summary = df.groupby("MaritalStatus")["Default"].agg(
    total_loans="count",
    defaults="sum",
    default_rate="mean"
).round(3)

print(summary)

               total_loans  defaults  default_rate
MaritalStatus                                     
Divorced             47646      5467         0.115
Married              47877      4616         0.096
Single               47917      5272         0.110


In [14]:
#Default rate analysis for Has Co Signer
summary = df.groupby("HasCoSigner")["Default"].agg(
    total_loans="count",
    defaults="sum",
    default_rate="mean"
).round(3)

print(summary)

             total_loans  defaults  default_rate
HasCoSigner                                     
No                 71794      8547         0.119
Yes                71646      6808         0.095


In [15]:
#Default rate analysis for Has Mortgage
summary = df.groupby("HasMortgage")["Default"].agg(
    total_loans="count",
    defaults="sum",
    default_rate="mean"
).round(3)

print(summary)

             total_loans  defaults  default_rate
HasMortgage                                     
No                 72020      8273         0.115
Yes                71420      7082         0.099


In [16]:
#Day 6
#Task 1- Creating Stress and stability flags

df["high_DTI_flag"] = (df["DTIRatio"]>= 0.5).astype(int)
df["low_credit_flag"] = (df["CreditScore"]<600).astype(int)
df["emp_risk_flag"] = df["EmploymentType"].isin(
    ["Unemployed", "Part-time", "Self-employed"]
).astype(int)
df["short_tenure_flag"] = (df["MonthsEmployed"]<24).astype(int)
df["has_cosigner_flag"] = (df["HasCoSigner"]=="Yes").astype(int)
df["has_asset_flag"] = (df["HasMortgage"]=="Yes").astype(int)

In [17]:
#Risk Score Calculation

df["Risk_Score"]= df["high_DTI_flag"]+df["low_credit_flag"]+df["emp_risk_flag"]+df["short_tenure_flag"]
Low_Risk_bucket =(df["Risk_Score"]<=1).sum()
Medium_Risk_bucket = (df["Risk_Score"]==2).sum()
High_Risk_bucket= (df["Risk_Score"]>2).sum()
print("Low Risk count:",Low_Risk_bucket)
print("Medium Risk count:",Medium_Risk_bucket)
print("High Risk count:",High_Risk_bucket)

Low Risk count: 46529
Medium Risk count: 57867
High Risk count: 39044


In [18]:
df.groupby("Risk_Score")["Default"].mean().round(3)


Risk_Score
0    0.074
1    0.086
2    0.106
3    0.130
4    0.177
Name: Default, dtype: float64

In [19]:
#Day 7
#Task 1 :Predicting defaults based on risk score (balanced)
df["Pred_Default_R3"]= (df["Risk_Score"]>=3).astype(int)
df["Pred_Default_R3"].sum()

np.int64(39044)

In [20]:
#Predicted defaults vs Actual default
print("No of Predicted default (risk score >= 3):", df["Pred_Default_R3"].sum(),"\nNo of Actual defaults in the dataset: ",df["Default"].sum())

No of Predicted default (risk score >= 3): 39044 
No of Actual defaults in the dataset:  15355


In [21]:
# finding true positive values
df["true_positives"]= ((df["Pred_Default_R3"]==1)&(df["Default"]==1)).astype(int)
TP = df["true_positives"].sum()
print("True positive: ",TP) 

True positive:  5337


In [22]:
# finding false positive values
df["false_positives"]= ((df["Pred_Default_R3"]==1)&(df["Default"]==0)).astype(int)
FP = df["false_positives"].sum()
print("false positive: ",FP)

false positive:  33707


In [23]:
# finding true negative values
df["true_negative"]= ((df["Pred_Default_R3"]==0)&(df["Default"]==0)).astype(int)
TN = df["true_negative"].sum()
print("true negative : ",TN)

true negative :  94378


In [24]:
# finding false negative values
df["false_negatives"]= ((df["Pred_Default_R3"]==0)&(df["Default"]==1)).astype(int)
FN =df["false_negatives"].sum()
print("false negative: ",FN)

false negative:  10018


In [25]:
#confusion matrix
confusion_matrix =TP+FP+TN+FN
confusion_matrix

np.int64(143440)

In [26]:
#actual defaulter  = TP + FN
actual_defaulter = TP +FN
actual_defaulter

np.int64(15355)

In [27]:
#Predicted defaulter = TP + FP
predicted_defaulter = TP + FP
predicted_defaulter

np.int64(39044)

In [28]:
#Recall(Default detection)
Recall = TP/(TP+FN)*100
Recall.round(2)

np.float64(34.76)

In [29]:
#Precision (Flag Accuracy)
Precsion_Flag = TP/(TP+FP)*100
Precsion_Flag.round(2)

np.float64(13.67)

In [30]:
#Approval rate
Approved_loan = (df["Pred_Default_R3"] == 0).sum()
Approval_rate= Approved_loan/total_no_of_loans*100
Approval_rate.round(2)

np.float64(72.78)

In [31]:
#Task 2 :Predicting defaults based on risk score (high safety)
df["Pred_Default_R2"] = (df["Risk_Score"] >= 2).astype(int)
df["Pred_Default_R2"].sum()

np.int64(96911)

In [32]:
#Predicted defaults vs Actual default
print("No of Predicted default (risk score >= 2):", df["Pred_Default_R2"].sum(),"\nNo of Actual defaults in the dataset: ",df["Default"].sum())

No of Predicted default (risk score >= 2): 96911 
No of Actual defaults in the dataset:  15355


In [33]:
# finding true positive values for R2
df["true_positives"]= ((df["Pred_Default_R2"]==1)&(df["Default"]==1)).astype(int)
TP = df["true_positives"].sum()
print("True positive: ",TP) 

True positive:  11465


In [34]:
# finding false positive values for R2
df["false_positives"]= ((df["Pred_Default_R2"]==1)&(df["Default"]==0)).astype(int)
FP = df["false_positives"].sum()
print("False positive: ",FP) 

False positive:  85446


In [35]:
# finding true negative values for R2
df["true_negative"]= ((df["Pred_Default_R2"]==0)&(df["Default"]==0)).astype(int)
TN = df["true_negative"].sum()
print("true negative : ",TN)

true negative :  42639


In [36]:
# finding false negative values
df["false_negatives"]= ((df["Pred_Default_R2"]==0)&(df["Default"]==1)).astype(int)
FN =df["false_negatives"].sum()
print("false negative: ",FN)

false negative:  3890


In [37]:
#confusion matrix for R2
confusion_matrix =TP+FP+TN+FN
confusion_matrix

np.int64(143440)

In [38]:
#actual defaulter  = TP + FN for R2
actual_defaulter = TP +FN
actual_defaulter

np.int64(15355)

In [39]:
#Predicted defaulter = TP + FP for R2
predicted_defaulter = TP + FP
predicted_defaulter

np.int64(96911)

In [40]:
#Recall(Default detection) for R2
Recall = TP/(TP+FN)*100
Recall.round(2)

np.float64(74.67)

In [41]:
#Precision (Flag Accuracy) for R2
Precsion_Flag = TP/(TP+FP)*100
Precsion_Flag.round(2)

np.float64(11.83)

In [42]:
#Approval rate for R2
approved_loans = (df["Pred_Default_R2"] == 0).sum()
Approval_rate= approved_loans/total_no_of_loans*100
Approval_rate.round(2)

np.float64(32.44)

In [43]:
#Task 3 :Predicting defaults based on risk score (high approval)
df["Pred_Default_R4"] = (df["Risk_Score"] >= 4).astype(int)
df["Pred_Default_R4"].sum()

np.int64(5535)

In [44]:
#Predicted defaults vs Actual default
print("No of Predicted default (risk score >= 4):", df["Pred_Default_R2"].sum(),"\nNo of Actual defaults in the dataset: ",df["Default"].sum())

No of Predicted default (risk score >= 4): 96911 
No of Actual defaults in the dataset:  15355


In [45]:
# finding true positive values for R4
df["true_positives"]= ((df["Pred_Default_R4"]==1)&(df["Default"]==1)).astype(int)
TP = df["true_positives"].sum()
print("True positive: ",TP) 

True positive:  980


In [46]:
# finding false positive values for R4
df["false_positives"]= ((df["Pred_Default_R4"]==1)&(df["Default"]==0)).astype(int)
FP = df["false_positives"].sum()
print("False positive: ",FP) 

False positive:  4555


In [47]:
# finding true negative values for R4
df["true_negative"]= ((df["Pred_Default_R4"]==0)&(df["Default"]==0)).astype(int)
TN = df["true_negative"].sum()
print("true negative : ",TN)

true negative :  123530


In [48]:
# finding false negative values for R4
df["false_negatives"]= ((df["Pred_Default_R4"]==0)&(df["Default"]==1)).astype(int)
FN =df["false_negatives"].sum()
print("false negative: ",FN)

false negative:  14375


In [49]:
#confusion matrix for R4
confusion_matrix = pd.DataFrame(
    [[TP, FP],
     [FN, TN]],
    index=["Pred_Default", "Pred_NonDefault"],
    columns=["Actual_Default", "Actual_NonDefault"]
)
confusion_matrix


Actual_Default  Actual_NonDefault
Pred_Default                980               4555
Pred_NonDefault           14375             123530

In [50]:
#actual defaulter  = TP + FN for R4
actual_defaulter = TP +FN
actual_defaulter

np.int64(15355)

In [51]:
#Predicted defaulter = TP + FP for R4
predicted_defaulter = TP + FP
predicted_defaulter

np.int64(5535)

In [52]:
#Recall(Default detection) for R4
Recall = TP/(TP+FN)*100
Recall.round(2)

np.float64(6.38)

In [53]:
#Precision (Flag Accuracy) for R4
Precsion_Flag = TP/(TP+FP)*100
Precsion_Flag.round(2)

np.float64(17.71)

In [54]:
#Approval rate for R4
approved_loans = (df["Pred_Default_R4"] == 0).sum()
Approval_rate= approved_loans/total_no_of_loans*100
Approval_rate.round(2)

np.float64(96.14)

Rule    |   Rejection  |   Approval    |   Default    |  Default    |   Business        
        |   Condition  |     Rate      |  Prevention  |  Caught     |   Interpretation
---------------------------------------------------------------------------------------
R2      |   Risk >=2   |     32.44     |     11465    |    74.67    | Aggressive risk control
--------------------------------------------------------------------------------------
R3      |   Risk >=3   |     72.78     |     5337     |    34.76    | Balanced
---------------------------------------------------------------------------------------
R4      |   Risk >=4   |     96.14     |     980      |     6.38    | Conservative

In [56]:
#Day-11 Model Preparation
keep = [
 "Age","Income","LoanAmount","CreditScore","MonthsEmployed","NumCreditLines",
 "LoanTerm","DTIRatio","Education","EmploymentType","MaritalStatus","HasMortgage",
 "HasDependents","LoanPurpose","HasCoSigner","has_cosigner_flag","has_asset_flag",
 "high_DTI_flag","low_credit_flag","emp_risk_flag","short_tenure_flag"
]

model_df = df[keep].copy()
model_df.to_csv("../Data/Processed/model_dataset.csv", index=False)
model_df.head()

Age  Income  LoanAmount  CreditScore  MonthsEmployed  NumCreditLines  \
0   64   73743      140354          300               0               2   
1   22   46351      151652          300               0               4   
2   56   42359      163692          300               0               3   
3   23   52705      218580          300               0               3   
4   58  131487       15557          301               0               4   

   LoanTerm  DTIRatio    Education EmploymentType  ... HasMortgage  \
0        12      0.24          PhD  Self-employed  ...         Yes   
1        36      0.51          PhD      Part-time  ...          No   
2        60      0.71          PhD      Part-time  ...          No   
3        60      0.54  High School     Unemployed  ...          No   
4        36      0.38     Master's      Part-time  ...          No   

  HasDependents LoanPurpose HasCoSigner has_cosigner_flag  has_asset_flag  \
0            No   Education         Yes                 1               1   
1           Yes       Other         Yes                 1               0   
2           Yes   Education          No                 0               0   
3           Yes    Business         Yes                 1               0   
4            No       Other         Yes                 1               0   

   high_DTI_flag  low_credit_flag  emp_risk_flag  short_tenure_flag  
0              0                1              1                  1  
1              1                1              1                  1  
2              1                1              1                  1  
3              1                1              1                  1  
4              0                1              1                  1  

[5 rows x 21 columns]